### PR TITLE
[C++ API parity] Soft Margin loss

### PR DESCRIPTION
**Pull request by CarMiranda**
In accordance with #25883, I added the `SoftMarginLoss` module and `soft_margin_loss` functional.

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 1]
```diff
@@ -39,5 +39,17 @@ struct TORCH_API CosineEmbeddingLossOptions {
...
+// ============================================================================
+
+/// Options for a soft margin loss functional and module.
+struct TORCH_API SoftMarginLossOptions {
+
+  /// Specifies the reduction to apply to the output: 'none' | 'mean' | 'sum'.
+  /// 'none': no reduction will be applied, 'mean': the sum of the output will
+  /// be divided by the number of elements in the output, 'sum': the output will
+  /// be summed. Default: 'mean'
+  TORCH_ARG(Reduction::Reduction, reduction) = Reduction::Mean;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: Similar to the `SmoothL1Loss` PR, we can add an implicit constructor and set default of `reduction` through the constructor.

File: test/cpp/api/functional.cpp [hunk 2]
```diff
@@ -73,6 +73,19 @@ TEST_F(FunctionalTest, CosineSimilarity) {
   ASSERT_TRUE(output.allclose(expected, 1e-04));
 }
 
+TEST_F(FunctionalTest, SoftMarginLoss) {
+  auto input = torch::tensor({2., 4., 1., 3.}, torch::requires_grad());
+  auto target = torch::tensor({-1., 1., 1., -1.}, torch::kFloat);
+  auto output =
+      F::soft_margin_loss(input, target, SoftMarginLossOptions());
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: We can remove `SoftMarginLossOptions()` in this line to test the "no options" use case :D

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 3]
```diff
@@ -49,5 +49,18 @@ Tensor CosineEmbeddingLossImpl::forward(
   return F::cosine_embedding_loss(input1, input2, target, options);
 }
 
+// ============================================================================
+
+SoftMarginLossImpl::SoftMarginLossImpl(
+    const torch::nn::SoftMarginLossOptions& options_) : options(options_) {}
+
+void SoftMarginLossImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::SoftMarginLoss(reduction=" << options.reduction() << ")";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: Similarly here, we can add a test for `pretty_print`, and as part of the test we can test that `SoftMarginLoss(/*reduction=*/Reduction::Mean)` (i.e. the implicit constructor of `SoftMarginLossOptions`) works :D

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 4]
```diff
@@ -49,5 +49,18 @@ Tensor CosineEmbeddingLossImpl::forward(
...
+}
+
+Tensor SoftMarginLossImpl::forward(const Tensor& input, const Tensor& target) {
+  return torch::soft_margin_loss(input, target, options.reduction());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/loss.cpp]: would be awesome to do:
```suggestion
  return F::soft_margin_loss(input, target, options);
```

File: torch/csrc/api/include/torch/nn/modules/loss.h [hunk 5]
```diff
@@ -90,5 +90,29 @@ struct TORCH_API CosineEmbeddingLossImpl : Module {
 /// PyTorch's module storage semantics.
 TORCH_MODULE(CosineEmbeddingLoss);
 
+
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/loss.h]: nit: we can remove the new line here

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 6]
```diff
@@ -26,6 +26,15 @@ inline Tensor cosine_embedding_loss(
       input1, input2, target, options.margin(), options.reduction());
 }
 
+inline Tensor soft_margin_loss(
+    const Tensor& input,
+    const Tensor& target,
+    const SoftMarginLossOptions& options) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/loss.h]: similarly regarding `options = {}` :D

**Review by yf225 — COMMENTED** [review]
Thanks so much for the fantastic work @CarMiranda! I left some comments.

**Review by yf225 — APPROVED** [review]
@CarMiranda Thanks a lot for the awesome work! I will merge it after CI passes :D

**Comment by yf225** [comment]
TODO: wait for https://github.com/pytorch/pytorch/pull/27422 to land first, and then use `at::Reduction` in this PR.
